src/Sema.zig

```diff
@@ -38109,3 +38109,273 @@ fn ptrType(sema: *Sema, info: InternPool.Key.PtrType) CompileError!Type {
+        .func = ip.extraFuncInstance(func_extra_index),
+    }, adapter);
+    if (gop.found_existing) {
+        // Hot path: undo the additions to our two arrays.
+        ip.items.len -= 4;
+        ip.extra.items.len = prev_extra_len;
+        return @enumFromInt(gop.index);
+    }
+
+    // Synchronize the map with items.
+    assert(!ip.map.getOrPutAssumeCapacityAdapted(InternPool.Key{ .error_union_type = .{
+        .error_set_type = error_set_type,
+        .payload_type = arg.bare_return_type,
+    } }, adapter).found_existing);
+    assert(!ip.map.getOrPutAssumeCapacityAdapted(InternPool.Key{
+        .inferred_error_set_type = func_index,
+    }, adapter).found_existing);
+    assert(!ip.map.getOrPutAssumeCapacityAdapted(InternPool.Key{
+        .func_type = ip.extraFuncType(func_type_extra_index),
+    }, adapter).found_existing);
+
+    return sema.finishFuncInstance(
+        generic_owner,
+        func_index,
+        func_extra_index,
+        arg.generation,
+        func_ty,
+        arg.section,
+    );
+}
+
+fn finishFuncInstance(
+    sema: *Sema,
+    generic_owner: InternPool.Index,
+    func_index: InternPool.Index,
+    func_extra_index: u32,
+    generation: u32,
+    func_ty: InternPool.Index,
+    section: InternPool.OptionalNullTerminatedString,
+) Allocator.Error!InternPool.Index {
+    const gpa = sema.gpa;
+    const ip = &sema.mod.intern_pool;
+
+    const fn_owner_decl = sema.mod.declPtr(ip.funcDeclOwner(generic_owner));
+    const decl_index = try sema.mod.createDecl(.{
+        .name = undefined,
+        .src_namespace = fn_owner_decl.src_namespace,
+        .src_node = fn_owner_decl.src_node,
+        .src_line = fn_owner_decl.src_line,
+        .has_tv = true,
+        .owns_tv = true,
+        .ty = @import("type.zig").Type.fromInterned(func_ty),
+        .val = @import("value.zig").Value.fromInterned(func_index),
+        .alignment = .none,
+        .@"linksection" = section,
+        .@"addrspace" = fn_owner_decl.@"addrspace",
+        .analysis = .complete,
+        .zir_decl_index = fn_owner_decl.zir_decl_index,
+        .src_scope = fn_owner_decl.src_scope,
+        .generation = generation,
+        .is_pub = fn_owner_decl.is_pub,
+        .is_exported = fn_owner_decl.is_exported,
+        .has_linksection_or_addrspace = fn_owner_decl.has_linksection_or_addrspace,
+        .has_align = fn_owner_decl.has_align,
+        .alive = true,
+        .kind = .anon,
+    });
+    errdefer sema.mod.destroyDecl(decl_index);
+
+    // Populate the owner_decl field which was left undefined until now.
+    ip.extra.items[
+        func_extra_index + std.meta.fieldIndex(InternPool.Tag.FuncInstance, "owner_decl").?
+    ] = @intFromEnum(decl_index);
+
+    // TODO: improve this name
+    const decl = sema.mod.declPtr(decl_index);
+    decl.name = try ip.getOrPutStringFmt(gpa, "{}__anon_{d}", .{
+        fn_owner_decl.name.fmt(ip), @intFromEnum(decl_index),
+    });
+
+    return func_index;
+}
```

Comment:
This really should all stay in InternPool. What was the motivation for moving it?